supporting wildcards in resource namespace matching

iamsgarg-ob opened the pull request:
## Explanation

<!--
In a couple sentences, explain why this PR is needed and what it addresses. This should be an explanation a non-developer user can understand and covers the "why" question. It should also clearly indicate whether this PR represents an addition, a change, or a fix of existing behavior. This explanation will be used to assist in the release note drafting process.

THIS IS MANDATORY.
-->
In a cluster policy, we list all resources in the namespace correctly, but when we attempt to match the desired resource namespace against the actual resource namespace - we do not consider wildcards . In the kyverno policy match clause namespaces can have * or ? wildcards. This PR adds code to do that match iff wildcards are present.

## Related issue
Closes #14212 

## Milestone of this PR
<!--

Add the milestone label by commenting `/milestone 1.2.3`.

-->

## Documentation (required for features)

My PR contains new or altered behavior to Kyverno. 
- [ ] I have sent the draft PR to add or update [the documentation](https://github.com/kyverno/website) and the link is:
  <!-- Uncomment to link to the PR -->
  <!-- https://github.com/kyverno/website/pull/123 -->

## What type of PR is this
 /kind bug

## Proposed Changes

<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->

### Proof Manifests

<!--
Read and follow the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) for more details first. This section is for pasting your YAML manifests (Kubernetes resources and Kyverno policies) and Kyverno CLI test manifests which allow maintainers to prove the intended functionality is achieved by your PR. Please use proper fenced code block formatting, for example:

# Kubernetes resource

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: roles-dictionary
  namespace: default
data:
  allowed-roles: "[\"cluster-admin\", \"cluster-operator\", \"tenant-admin\"]"
```

# Kyverno CLI test manifest (please see docs for latest manifest format at https://kyverno.io/docs/kyverno-cli/). See kyverno/policies for complete examples of all related test files.

```yaml
name: prepend-image-registry
policies:
  - prepend_image_registry.yaml
resources:
  - resource.yaml
variables: values.yaml
results:
  - policy: prepend-registry
    rule: prepend-registry-containers
    resource: mypod
    # if mutate rule
    patchedResource: patchedResource01.yaml
    kind: Pod
    result: pass
```
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [X] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [X] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [X] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added and my PR doesn't contain that functionality.

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

iamsgarg-ob commented on pkg/policy/utils.go:
```diff
@@ -26,12 +29,30 @@ func resourceMatches(match kyvernov1.ResourceDescription, res unstructured.Unstr
 	return true
 }
 
+func containsIncludingWildcards(slice []string, item string) bool {
```
created a separate containsIncludingWildcards since the existing `contains` function pretty much does an "equality" check only - and is called in different places. The codepath invoking `resourceMatches` is a singular one from the policy controller hence is safe.

realshuting replied:
We can reuse wildcard lib, see https://github.com/kyverno/kyverno/blob/main/ext/wildcard/match.go#L7.

iamsgarg-ob commented:
> Thanks for the PR @iamsgarg-ob !
> 
> Please see my comments below:
> - Can you please add Chainsaw tests to the [pool](https://github.com/kyverno/kyverno/tree/main/test/conformance/chainsaw/generate/clusterpolicy/cornercases)?
> - A couple of CI tests failed, can you look into those?


Hey @realshuting so the checks all pass in my local fork after including your suggestions. 

Could you please guide on what the milestone label should be?

realshuting commented on pkg/policy/utils.go:
```diff
@@ -26,12 +28,25 @@ func resourceMatches(match kyvernov1.ResourceDescription, res unstructured.Unstr
+		}
+	}
+	return false
+}
```
Can we simplify it to - 
```
func containsIncludingWildcards(slice []string, item string) bool {
	for _, s := range slice {
		if wildcard.Match(s, item) {
			return true
		}
	}
	return false
}
```

iamsgarg-ob commented:
@realshuting I have verified the original policy attached to the bug report in my local kind cluster. 

* Created 3 namespaces tn-1,tn-2,tn-3 and added serviceAccounts in them WITHOUT imagePullSecrets.
* Created a 4th namespace tn-4 . Added a ServiceAccount WITH a imagePullSecret
* Created the imagePullSecret in each namespace
* Applied the Cpol as mentioned in the original bug report
* Checked the ServiceAccounts  in tn-1, tn-2, tn-3 - they had the imagPullSecret correctly populated . No change was observed in tn-4 ( as expected ) 

Log Lines from Background Controller: 

```
│ 2025-11-20T05:40:17Z -3 github.com/kyverno/kyverno/pkg/engine/context/context.go:274 > Adding service account logger=context service account name= service account namespace= v=4                                                          │
│ 2025-11-20T05:40:17Z -2 github.com/kyverno/kyverno/pkg/engine/handlers/mutation/mutate_existing.go:66 > processing mutate rule logger=engine.mutate new.kind=ServiceAccount new.name=demo-sa new.namespace=tn-foo old.kind=ServiceAccount  │
│ old.name=demo-sa old.namespace=tn-foo policy.apply=All policy.name=patch-tenant-serviceaccounts-with-acr-pull-secret policy.namespace= rule.name=add-acr-pull-secret v=3                                                                   │
│ 2025-11-20T05:40:17Z -2 github.com/kyverno/kyverno/pkg/engine/variables/vars.go:359 > variable substituted logger=engine.mutate new.kind=ServiceAccount new.name=demo-sa new.namespace=tn-foo old.kind=ServiceAccount old.name=demo-sa old │
│ .namespace=tn-foo path=/namespace policy.apply=All policy.name=patch-tenant-serviceaccounts-with-acr-pull-secret policy.namespace= rule.name=add-acr-pull-secret v=3 value=tn-foo variable="{{ request.object.metadata.namespace }}"       │
│ 2025-11-20T05:40:17Z -2 github.com/kyverno/kyverno/pkg/engine/variables/vars.go:359 > variable substituted logger=engine.mutate new.kind=ServiceAccount new.name=demo-sa new.namespace=tn-foo old.kind=ServiceAccount old.name=demo-sa old │
│ .namespace=tn-foo path=/name policy.apply=All policy.name=patch-tenant-serviceaccounts-with-acr-pull-secret policy.namespace= rule.name=add-acr-pull-secret v=3 value=demo-sa variable="{{ request.object.metadata.name }}"                │
│ 2025-11-20T05:40:17Z -2 github.com/kyverno/kyverno/pkg/engine/variables/vars.go:359 > variable substituted logger=engine.mutate new.kind=ServiceAccount new.name=demo-sa new.namespace=tn-foo old.kind=ServiceAccount old.name=demo-sa old │
│ .namespace=tn-foo path= policy.apply=All policy.name=patch-tenant-serviceaccounts-with-acr-pull-secret policy.namespace= rule.name=add-acr-pull-secret v=3 value=false variable="{{ contains(keys(target), 'imagePullSecrets') }}"         │
│ 2025-11-20T05:40:17Z -2 github.com/kyverno/kyverno/pkg/engine/variables/vars.go:359 > variable substituted logger=engine.mutate new.kind=ServiceAccount new.name=demo-sa new.namespace=tn-foo old.kind=ServiceAccount old.name=demo-sa old │
│ .namespace=tn-foo path=/mutate/targets/0/namespace policy.apply=All policy.name=patch-tenant-serviceaccounts-with-acr-pull-secret policy.namespace= rule.name=add-acr-pull-secret v=3 value=tn-foo variable="{{ request.object.metadata.na │
│ mespace }}"
```

Test #2:
* Deleted the imagePullSecret from an existing ServiceAccount. 
* The imagePullSecret was added back by the background controller.
